File: csharp/1499.max-value-of-equation.cs
/*
 * @lc app=leetcode id=1499 lang=csharp
 *
 * [1499] Max Value of Equation
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int FindMaxValueOfEquation(int[][] points, int k) {
        var list = new LinkedList<int[]>();
        int ret = int.MinValue;
        foreach (var p in points) {
            while (list.Count > 0 && list.First.Value[0] < p[0] - k) {
                list.RemoveFirst();
            }
            if (list.Count > 0) {
                ret = Math.Max(ret, p[1] + list.First.Value[1] + p[0] - list.First.Value[0]);
                int diff = p[1] - p[0];
                while (list.Count > 0 && list.Last.Value[1] - list.Last.Value[0] < diff) {
                    list.RemoveLast();
                }
            }
            list.AddLast(p);
        }
        return ret;
    }
}
// @lc code=end

File: java/1499.max-value-of-equation.java
/*
 * @lc app=leetcode id=1499 lang=java
 *
 * [1499] Max Value of Equation
 */

import java.util.ArrayDeque;
import java.util.Deque;

// @lc code=start
class Solution {
    public int findMaxValueOfEquation(int[][] points, int k) {
        Deque<int[]> deque = new ArrayDeque<>();
        int ret = Integer.MIN_VALUE;
        for (int[] point : points) {
            while (!deque.isEmpty() && deque.getFirst()[0] < point[0] - k) {
                deque.removeFirst();
            }
            if (!deque.isEmpty()) {
                ret = Math.max(ret, point[1] + deque.getFirst()[1] + point[0] - deque.getFirst()[0]);
            }
            int diff = point[1] - point[0];
            while (!deque.isEmpty() && deque.getLast()[1] - deque.getLast()[0] < diff) {
                deque.removeLast();
            }
            deque.addLast(point);
        }
        return ret;
    }
}
// @lc code=end

